File: News_Service_Jooq/src/main/java/com/dunice/projectNews/interceptor/CustomInterceptor.java
package com.dunice.projectNews.interceptor;

import com.dunice.projectNews.models.RequestLog;
import com.dunice.projectNews.repository.RequestLogRepository;
import com.dunice.projectNews.util.AppConstants;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;
import org.springframework.web.servlet.HandlerInterceptor;
import java.time.LocalDateTime;

@RequiredArgsConstructor
@Component
public class CustomInterceptor implements HandlerInterceptor {

    private final RequestLogRepository requestLogRepository;

    @Override
    public void afterCompletion(HttpServletRequest request,
                                HttpServletResponse response,
                                Object handler, Exception ex) throws Exception {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        String username = (authentication != null) ? authentication.getName() : AppConstants.NO_AUTHORIZATION;
        String method = request.getMethod();
        String url = request.getRequestURL().toString();
        int status = response.getStatus();
        String error = (String) request.getAttribute(AppConstants.ERRORS);

        if (error == null) {
            error = AppConstants.MESSAGE_STATUS_OK;
        }
        RequestLog log = new RequestLog(
                LocalDateTime.now(),
                method,
                url,
                status,
                error,
                username
        );
        requestLogRepository.save(log);
    }
}

File: News_Service_Kotlin/src/main/kotlin/com/dunice/advancedServiceKotlin/interceptor/CustomInterceptor.kt
package com.dunice.advancedServiceKotlin.interceptor

import com.dunice.advancedServiceKotlin.models.RequestLog
import com.dunice.advancedServiceKotlin.repository.RequestLogRepository
import com.dunice.advancedServiceKotlin.util.AppConstants
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component
import org.springframework.web.servlet.HandlerInterceptor
import java.time.LocalDateTime

@Component
class CustomInterceptor(private val requestLogRepository: RequestLogRepository) : HandlerInterceptor {
    @Throws(Exception::class)
    override fun afterCompletion(
        request: HttpServletRequest,
        response: HttpServletResponse,
        handler: Any, ex: Exception?
    ) {
        val authentication = SecurityContextHolder.getContext().authentication
        val username = if ((authentication != null)) authentication.name else AppConstants.NO_AUTHORIZATION
        val method = request.method
        val url = request.requestURL.toString()
        val status = response.status
        var error = request.getAttribute(AppConstants.ERRORS) as String

        if (error == null) {
            error = AppConstants.MESSAGE_STATUS_OK
        }
        val log = RequestLog(
            LocalDateTime.now(),
            method,
            url,
            status,
            error,
            username
        )
        requestLogRepository.save(log)
    }
}
